src/PromTracker.mo

```diff
@@ -0,0 +1,291 @@
+import AssocList "mo:base/AssocList";
+import Cycles "mo:base/ExperimentalCycles";
+import Int "mo:base/Int";
+import Iter "mo:base/Iter";
+import Nat "mo:base/Nat";
+import Nat64 "mo:base/Nat64";
+import Prim "mo:prim";
+import StableMemory "mo:base/ExperimentalStableMemory";
+import Time "mo:base/Time";
+import Text "mo:base/Text";
+
+import Vec "mo:vector";
+
+module {
+
+  type StableDataItem = { #counter : Nat; #gauge : (Nat, Nat) };
+  public type StableData = AssocList.AssocList<Text, StableDataItem>;
+
+  /// An access interface for pull value
+  public type PullValueInterface = {
+    value : () -> Nat;
+    remove : () -> ();
+  };
+  /// An access interface for counter value
+  public type CounterInterface = {
+    value : () -> Nat;
+    set : (x : Nat) -> ();
+    add : (x : Nat) -> ();
+    remove : () -> ();
+  };
+  /// An access interface for gauge value
+  public type GaugeInterface = {
+    value : () -> Nat;
+    update : (x : Nat) -> ();
+    remove : () -> ();
+  };
+
+  /// Value tracker, designed specifically to use as source for Prometheus.
+  ///
+  /// Example:
+  /// ```motoko
+  /// let tracker = PromTracker.PromTracker(65); // 65 seconds is the recommended interval if prometheus pulls stats with interval 60 seconds
+  /// ....
+  /// let successfulHeartbeats = tracker.addCounter("successful_heartbeats", true);
+  /// let failedHeartbeats = tracker.addCounter("failed_heartbeats", true);
+  /// let heartbeats = tracker.addPullValue("heartbeats", func() = successfulHeartbeats.value() + failedHeartbeats.value());
+  /// let heartbeatDuration = tracker.addGauge("heartbeat_duration", true);
+  /// ....
+  /// // update values from your code
+  /// successfulHeartbeats.add(2);
+  /// failedHeartbeats.add(1);
+  /// heartbeatDuration.update(10);
+  /// heartbeatDuration.update(18);
+  /// heartbeatDuration.update(14);
+  /// ....
+  /// // get prometheus metrics:
+  /// let text = tracker.renderStats();
+  /// ```
+  ///
+  /// Expected output is:
+  /// ```
+  /// successful_heartbeats{} 2 1698842860811
+  /// failed_heartbeats{} 1 1698842860811
+  /// heartbeats{} 3 1698842860811
+  /// heartbeat_duration_sum{} 42 1698842860811
+  /// heartbeat_duration_count{} 3 1698842860811
+  /// heartbeat_duration_high_watermark{} 18 1698842860811
+  /// heartbeat_duration_low_watermark{} 10 1698842860811
+  /// ```
+  public class PromTracker(watermarkResetIntervalSeconds : Nat) {
+
+    type IValue = {
+      prefix : Text;
+      dump : () -> [(Text, Nat)];
+      share : () -> ?StableDataItem;
+      unshare : (StableDataItem) -> ();
+    };
+
+    let values_ : Vec.Vector<?IValue> = Vec.new<?IValue>();
```

Comment:
Can we rename this to `values`? I don't see a competing function for that name.

And it might be nicer for subsequent code to use the class version of Vector. `add, size, put, vals` are all class members.

Reply:
Done